synapse/lib/multislabseqn.py

```diff
@@ -0,0 +1,420 @@
+    '''
+    An append-optimized sequence of byte blobs stored across multiple slabs for fast rotating/culling
+    '''
+
+    async def __anit__(self, dirn, slabopts=None):
+
+        await s_base.Base.__anit__(self)
+
+        self.IDX_PER_SLAB = int(os.environ.get('SYN_MULTISLAB_MAX_INDEX', DEF_IDX_PER_SLAB))
+
+        self.offsevents: List[Tuple[int, int, asyncio.Event]] = [] # as a heap
+        self._waitcounter = 0
+
+        self.dirn: str = dirn
+        s_common.gendir(self.dirn)
+        self.slabopts: Dict[Any] = {} if slabopts is None else slabopts
+
+        # The last/current slab
+        self.tailslab: Optional[s_lmdbslab.Slab] = None
+        self.tailseqn: Optional[s_slabseqn.SlabSeqn] = None
+
+        # The most recently accessed slab/seqn that isn't the tail
+        self._cacheslab: Optional[s_lmdbslab.Slab] = None
+        self._cacheseqn: Optional[s_slabseqn.SlabSeqn] = None
+        self._cacheridx = None
+
+        # A startidx -> (Slab, Seqn) dict for all open Slabs, so we don't accidentally open the same Slab twice
+        self._openslabs: Dict[int, Tuple[s_lmdbslab.Slab, s_slabseqn.SlabSeqn]] = {}
+
+        # Lock to avoid an open race
+        self._openlock = asyncio.Lock()
+
+        await self._discoverRanges()
+
+        async def fini():
+            for slab, _ in list(self._openslabs.values()):
+                # We incref the slabs, so might have to fini multiple times
+                count = 1
+                while count:
+                    count = await slab.fini()
+
+        self.onfini(fini)
+
+    def __repr__(self):
+        return f'MultiSlabSeqn: {self.dirn!r}'
+
+    @staticmethod
+    def _getFirstIndx(slab) -> Optional[int]:
+        db = slab.initdb('info')
+        bytz = slab.get(b'firstindx', db=db)
+        if bytz is None:
+            return 0
+        return s_common.int64un(bytz)
+
+    @staticmethod
+    def _setFirstIndx(slab, indx) -> bool:
+        db = slab.initdb('info')
+        return slab.put(b'firstindx', s_common.int64en(indx), db=db)
+
+    async def _discoverRanges(self):
+        '''
+        Go through the slabs and get the starting indices of the sequence in each slab
+        '''
+        fnstartidx = 0
+        lastidx = None
+        self._ranges: List[int] = []  # Starting offsets of all the slabs in order
+        self.firstindx = 0  # persistently-stored indicator of lowest index
+        self.indx = 0  # The next place an add() will go
+        lowindx = None
+
+        # Make sure the files are in order
+
+        for fn in sorted(s_common.listdir(self.dirn, glob='*seqn' + '[abcdef01234567890]' * 16 + '.lmdb')):
+
+            if not os.path.isdir(fn):
+                logger.warning(f'Found a non-directory {fn} where a directory should be')
+                continue
+            match = seqnslabre.match(os.path.basename(fn))
+            assert match
+
+            newstartidx = int.from_bytes(s_common.uhex(match.group(1)), 'big')
```

Comment:
Update these to `int(match.group(1), 16)` for simplicity

Reply:
ok